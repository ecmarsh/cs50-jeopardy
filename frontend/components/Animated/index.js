import Ease from './_Ease';
import Fade from './_Fade';
import Grow from './_Grow';
import Loud from './_Loud';
import Shift from './_Shift';
import Slide from './_Slide';
import Tally from './_Tally';
import Tree from './_Tree';
import Unfocus from './_Unfocus';

export { Ease, Fade, Grow, Loud, Shift, Slide, Tally, Tree, Unfocus };
