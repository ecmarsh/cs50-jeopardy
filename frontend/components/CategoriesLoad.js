import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { categoriesPerPage } from '../config';
import CategoryQuestions from './CategoryQuestions';
import { Heading } from './Views';
import Pagination from './Pagination';
import Error from './ErrorMessage';
import Main from './styles/Main';
import { Divider } from './styles/Utilities';

const CATEGORIES_QUERY = gql`
  query CATEGORIES_QUERY($gameName: String, $first: Int, $skip: Int) {
    categories(
      where: { game: { name: $gameName } }
      first: $first
      skip: $skip
    ) {
      id
      name
    }
  }
`;

class LoadCategories extends Component {
  state = {};

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { game, round, headingProps } = this.props;
    return (
      <Query
        query={CATEGORIES_QUERY}
        variables={{
          gameName: game,
          first: categoriesPerPage,
          skip: categoriesPerPage * (round - 1),
        }}
      >
        {({ data, loading, error }) => {
          if (loading) return <p>Loading...</p>;
          if (error) return <Error error={error} />;
          if (!data.categories) {
            return <Heading game={game} {...headingProps} />;
          }
          return (
            <Main>
              <Heading game={game} {...headingProps} />
              <Divider style={{ margin: '0 auto' }} />
              <Pagination gameName={game} round={round} />
              <div className="canvas">
                {data.categories.map(category => (
                  <div
                    key={'category' + category.id}
                    className="column-wrapper"
                  >
                    <ul>
                      <li>{category.name}</li>
                      <CategoryQuestions
                        categoryId={category.id}
                        gameName={game}
                        round={round}
                      />
                    </ul>
                  </div>
                ))}
              </div>
            </Main>
          );
        }}
      </Query>
    );
  }
}

export default LoadCategories;
export { CATEGORIES_QUERY };
