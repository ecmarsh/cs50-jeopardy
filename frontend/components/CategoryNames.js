import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import * as R from 'ramda';
import formatRound from '../lib/formatRound';
import Error from './ErrorMessage';
import Form from './styles/Form';
import Button from './styles/Button';
import { TextMuted, Divider } from './styles/Utilities';

const UPDATE_CATEGORY_MUTATION = gql`
  mutation UPDATE_CATEGORY_MUTATION($categoryId: ID!, $categoryName: String) {
    updateCategory(id: $categoryId, name: $categoryName) {
      id
      name
    }
  }
`;

class CategoryNames extends Component {
  static propTypes = {
    categories: PropTypes.array.isRequired,
  };

  state = {
    categories: [],
    called: false,
  };

  componentDidMount() {
    const { categories } = this.props;
    this.setState({ categories });
  }

  handleChange = e => {
    e.preventDefault();
    const { id, value } = e.target;
    const categories = [...this.state.categories];
    const idx = R.findIndex(R.propEq('id', id))(categories);
    categories[idx].name = value;
    categories[idx].updated = true; // flag to include in mutation
    this.setState({ categories });
  };

  handleSubmit = (e, updateCategoryMutation) => {
    // Stop default form submission
    e.preventDefault();
    // Get list of changed categories to update
    const updatedCategories = R.filter(R.has('updated'), this.state.categories);
    // Update category mutation
    const doMutation = async category => {
      const { id, name } = category;
      const res = await updateCategoryMutation({
        variables: { categoryId: id, categoryName: name },
      });
    };
    // Invoke mutation on each updated category
    R.forEach(doMutation, updatedCategories);
    // Called to flash saved message
    this.setState({ called: true });

    setTimeout(this.flashUpdatedConfirmation, 2000);
  };

  flashUpdatedConfirmation = () => {
    this.setState({ called: false });
  };

  render() {
    const { categories, called } = this.state;
    return (
      <Mutation mutation={UPDATE_CATEGORY_MUTATION}>
        {(updateCategory, { loading, error }) => {
          return (
            <Form
              data-test="category-names-form"
              className="noBg"
              onSubmit={e => this.handleSubmit(e, updateCategory)}
            >
              <h2>Category Names</h2>
              <Error error={error} />
              <fieldset disabled={loading} aria-busy={loading}>
                {categories.map((category, i) => (
                  <div key={category.id}>
                    {i === 0 && <h4>{formatRound(1)}</h4>}
                    {i === 5 && <h4>{formatRound(2)}</h4>}
                    {i === 10 && <h4>{formatRound(3)}</h4>}
                    <label
                      htmlFor={`category-${i}`}
                      style={{
                        flexDirection: 'row',
                        width: '100%',
                      }}
                    >
                      <TextMuted
                        style={{ textAlign: 'center', flexBasis: '25%' }}
                      >
                        CATEGORY {i + 1}:
                      </TextMuted>
                      <input
                        type="text"
                        name={`category-${i}`}
                        id={category.id}
                        value={category.name}
                        onChange={e => this.handleChange(e)}
                        style={{
                          flexBasis: '60%',
                          width: 'auto',
                        }}
                      />
                    </label>
                  </div>
                ))}
                <div style={{ margin: '0 auto' }}>
                  <TextMuted
                    data-test="category-names-save-confirmation"
                    hidden={!called}
                  >
                    Category Names Saved!
                  </TextMuted>
                  <Button
                    primary
                    type="submit"
                    style={{ margin: '1rem auto 2rem' }}
                  >
                    Save Category Names
                  </Button>
                  <Divider />
                  {this.props.children}
                </div>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }
}

export default CategoryNames;
export { UPDATE_CATEGORY_MUTATION };
