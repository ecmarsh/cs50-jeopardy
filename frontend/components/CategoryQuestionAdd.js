import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import DropTarget from './styles/DropTarget';
import Error from './ErrorMessage';
import RemoveFromCategory from './CategoryQuestionRemove';
import { GAME_QUESTIONS_QUERY } from './QuestionsList';
import { CATEGORY_QUESTIONS_QUERY } from './CategoryQuestions';
import { GAME_SUMMARY_QUERY } from './GameSummary';
import * as R from 'ramda';

const ADD_TO_CATEGORY_MUTATION = gql`
  mutation ADD_TO_CATEGORY_MUTATION($id: ID!, $questionId: ID!) {
    updateCategoryQuestion(id: $id, questionId: $questionId) {
      id
      question {
        categoryQuestions {
          id
        }
      }
    }
  }
`;

class AddToCategory extends Component {
  static propTypes = {
    loaded: PropTypes.bool,
    question: PropTypes.object,
    categoryQuestionId: PropTypes.string.isRequired,
    categoryId: PropTypes.string.isRequired,
    difficulty: PropTypes.number.isRequired,
    categoryName: PropTypes.string.isRequired,
    gameName: PropTypes.string.isRequired,
  };

  state = {
    questionId: String(),
    question: String(),
  };

  componentDidMount = () => {
    if (this.props.question) {
      const { id, question } = this.props.question;
      this.setState({
        questionId: id,
        question,
      });
    }
  };

  handleDragOver = e => {
    e.preventDefault();
    e.persist(); // don't pool the events
    e.target.style.transform = 'scale(1.2)';
    e.dataTransfer.dropEffect = 'move';
  };

  handleDragLeave = e => {
    e.target.style.transform = 'scale(1)';
  };

  handleDrop = async (e, categoryMutation) => {
    e.persist();
    e.target.style.transform = 'scale(1)';
    e.dataTransfer.dropEffect = 'move';

    const { categoryId, categoryQuestionId, difficulty, gameName } = this.props,
      questionId = e.dataTransfer.getData('id'),
      question = e.dataTransfer.getData('question'),
      answer = e.dataTransfer.getData('answer');

    this.setState({ questionId, question });

    const res = await categoryMutation({
      variables: { id: categoryQuestionId, questionId },
      update: (cache, { data: { updateCategoryQuestion } }) => {
        /* Update category question with added question */
        const categoryQuestionsQuery = {
          query: CATEGORY_QUESTIONS_QUERY,
          variables: { categoryId },
        };
        // Read the data
        const { categoryQuestions } = cache.readQuery(categoryQuestionsQuery);
        // Add the question
        categoryQuestions[difficulty - 1].question = {
          answer,
          id: questionId,
          question,
          __typename: 'Question',
        };
        // Write back the cache
        cache.writeQuery({
          data: { categoryQuestions },
          ...categoryQuestionsQuery,
        });

        /* Update question to know it's in a category now */
        const gameQuestionsQuery = {
          query: GAME_QUESTIONS_QUERY,
          variables: { gameName },
        };
        // Read data to questions variable
        const { questions } = cache.readQuery(gameQuestionsQuery);
        // Add new question into matching question's category list
        const idx = R.findIndex(R.propEq('id', questionId), questions);
        const questionToAdd = {
          id: updateCategoryQuestion.id,
          __typename: 'CategoryQuestion',
          category: {
            id: categoryId,
            __typename: 'Category',
          },
        };
        questions[idx].categoryQuestions.push(questionToAdd);
        // Write back cache with updates
        gameQuestionsQuery.data = { questions };
        //console.log(gameQuestionsQuery);
        cache.writeQuery(gameQuestionsQuery);
      },
    });
  };

  render() {
    const {
      loaded,
      question,
      categoryQuestionId,
      categoryId,
      difficulty,
      categoryName,
      gameName,
    } = this.props;

    return (
      <Mutation
        mutation={ADD_TO_CATEGORY_MUTATION}
        refetchQueries={[
          {
            query: GAME_SUMMARY_QUERY,
            variables: { gameName },
          },
        ]}
      >
        {(updateCategoryQuestion, { loading, error }) => {
          if (loading) return <DropTarget>Adding..</DropTarget>;
          if (error) return <Error error={error} />;
          return (
            <DropTarget
              id={`${categoryName}-${categoryId}`}
              onDragOver={!loaded ? this.handleDragOver : undefined}
              onDragLeave={!loaded ? this.handleDragLeave : undefined}
              onDrop={e => this.handleDrop(e, updateCategoryQuestion)}
              className={loaded ? '--isLoaded' : '--isNotLoaded'}
            >
              {loaded ? <div>{question.question}</div> : difficulty}
              {loaded && (
                <RemoveFromCategory
                  categoryQuestionId={categoryQuestionId}
                  categoryId={categoryId}
                  gameName={gameName}
                  difficulty={difficulty}
                  questionId={this.state.questionId}
                />
              )}
            </DropTarget>
          );
        }}
      </Mutation>
    );
  }
}

export default AddToCategory;
export { ADD_TO_CATEGORY_MUTATION };
