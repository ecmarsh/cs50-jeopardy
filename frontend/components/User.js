import React from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';

const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      email
      name
      permissions
    }
  }
`;

// User render prop
const User = props => (
  <Query {...props} query={CURRENT_USER_QUERY}>
    {payload => props.children(payload)}
  </Query>
);
User.propTypes = {
  children: PropTypes.func.isRequired,
};

// User hoc
function withUser(WrappedComponent) {
  function ComponentWithData(props) {
    return (
      <User>{({ data }) => <WrappedComponent me={data.me} {...props} />}</User>
    );
  }
  return ComponentWithData;
}

export default User;
export { CURRENT_USER_QUERY, withUser };
