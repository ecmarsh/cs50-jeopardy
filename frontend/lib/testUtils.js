import casual from 'casual';
import mappingArray from './mappingArray';

// seed it for repeatable random sequence
casual.seed(123);

const fakeQuestion = () => ({
  id: 'abc123',
  question: 'Time complexity of merge sort?',
  answer: 'n log n',
  user: {
    id: casual.uuid,
    name: 'User Name',
    __typename: 'User',
  },
  createdAt: '2019-02 - 26T19: 24: 16.000Z',
  game: {
    id: 'game123',
    name: 'GameForGeekz',
    user: {
      id: casual.uuid,
      __typename: 'User',
    },
    isPublic: casual.coin_flip,
    __typename: 'Game',
  },
  categoryQuestions: [
    {
      id: 'catQuest123',
      category: { id: 'cat123', __typename: 'Category' },
      __typename: '__CategoryQuestion',
    },
  ],
  __typename: 'Question',
});

const fakeUser = () => ({
  __typename: 'User',
  id: 'userABC123',
  name: casual.name,
  email: casual.email,
  permissions: ['USER'],
});

const fakeAdmin = () => ({
  __typename: 'User',
  id: 'adminABC123',
  name: casual.name,
  email: casual.email,
  permission: ['ADMIN', 'USER'],
  games: [],
});

const fakeCategoryQuestion = (id = 'catQuest123', difficulty = 1) => ({
  id,
  difficulty,
  isDouble: false,
  answered: casual.coin_flip,
  question: fakeQuestion(),
  category: { id: 'cat123', name: casual.word, __typename: 'Category' },
  game: { id: 'g@m3', name: 'Test Game', __typename: 'Game' },
  __typename: 'CategoryQuestion',
});

const fillCategoryQuestions = (n = 5) =>
  mappingArray(n).map(i => fakeCategoryQuestion(`catQuest-${i}`, i));

const fakeCategory = (id = 'cat123') => ({
  id,
  name: casual.word,
  game: { id: 'g@m3', name: 'Test Game' },
  categoryQuestions: fillCategoryQuestions(),
  __typename: 'Category',
});

const fillCategories = (n = 11) =>
  mappingArray(n).map(i => fakeCategory(`cat-${i}`));

const fakeTeam = id => ({
  __typename: 'Team',
  id: `t${id}`,
  name: `Team${id}`,
  score: 1000,
  game: { id: 'g@m3', name: 'Test Game' },
});

const fakeGameConfig = () => ({
  id: 'config123',
  hasDoubleJeopardy: false,
  hasRoundTimer: false,
  roundTime: 0,
  finalTime: 0,
  __typename: 'GameConfig',
});

const fakeGame = () => ({
  __typename: 'Game',
  id: 'g@m3',
  name: 'Test Game',
  createdAt: '2019-03 - 06T19: 12: 16.000Z',
  isPublic: casual.coin_flip,
  user: fakeUser(),
  teams: [fakeTeam(1), fakeTeam(2)],
  categories: fillCategories(),
  questions: [fakeQuestion()],
  config: fakeGameConfig(),
});

const fakeStudyGame = (id = 'SG123') => ({
  __typename: 'StudyGame',
  id,
  answeredQuestions: [fakeCategoryQuestion(), fakeCategoryQuestion()],
  score: 1000,
  game: fakeGame(),
  user: fakeUser(),
  createdAt: '2019-04 - 25T19: 10: 16.000Z',
});

const fakeGameSummary = () => ({
  questionsConnection: {
    aggregate: {
      count: casual.double + 1,
      __typename: 'aggregate',
    },
    __typename: 'questionsConnection',
  },
  categoryQuestionsConnection: {
    aggregate: {
      count: casual.double,
      __typename: 'aggregate',
    },
    __typename: 'categoryQuestionsConnection',
  },
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeQuestion,
  fakeUser,
  fakeAdmin,
  fakeCategory,
  fakeCategoryQuestion,
  fakeTeam,
  fakeGame,
  fakeStudyGame,
  fakeGameSummary,
  fakeGameConfig,
};
