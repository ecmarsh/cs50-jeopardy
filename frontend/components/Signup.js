import React, { PureComponent } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Link from 'next/link';
import { CURRENT_USER_QUERY } from './User';
import { withUser } from './User';
import Error from './ErrorMessage';
import Button from './styles/Button';
import { Alert } from './Views';
import Form, { Extension } from './styles/Form';
import { TextMuted } from './styles/Utilities';

const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $name: String!
    $email: String!
    $password: String!
    $access: String!
  ) {
    signup(email: $email, password: $password, name: $name, access: $access) {
      id
      name
      email
    }
  }
`;

const clearedState = {
  name: '',
  email: '',
  password: '',
  access: 'basic',
};
class Signup extends PureComponent {
  state = clearedState;

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { name, email, password, access } = this.state;
    if (this.props.me) {
      const message = `Looks like you are already a user! Please sign out first if you'd like to create another account.`;
      return <Alert message={message} />;
    }
    return (
      <Mutation
        mutation={SIGNUP_MUTATION}
        variables={this.state}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {(signup, { loading, error }) => (
          <Form
            method="POST"
            onSubmit={async e => {
              e.preventDefault();
              const res = await signup();
              //this.setState(clearedState);
              const pathname = access === `basic` ? `/submit` : `/`;
              Router.push({ pathname });
            }}
          >
            <Error error={error} />
            <h2>Create Account</h2>
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="name">
                <input
                  type="text"
                  name="name"
                  id="name"
                  placeholder="Name"
                  required
                  value={name}
                  minLength="2"
                  onChange={this.handleChange}
                />
                <TextMuted>Name</TextMuted>
              </label>
              <label htmlFor="email">
                <input
                  type="email"
                  name="email"
                  id="email"
                  placeholder="Email"
                  required
                  value={email}
                  onChange={this.handleChange}
                  validate="true"
                />
                <TextMuted>Email</TextMuted>
              </label>
              <label htmlFor="password">
                <input
                  type="password"
                  name="password"
                  id="password"
                  placeholder="Password"
                  required
                  value={password}
                  onChange={this.handleChange}
                />
                <TextMuted>Password</TextMuted>
              </label>
              <label htmlFor="access">
                <select
                  name="access"
                  id="access"
                  required
                  defaultValue={access}
                  onChange={this.handleChange}
                  style={{ minHeight: '3.5rem', minWidth: '27rem' }}
                >
                  <option value="basic">Basic</option>
                  <option value="moderator">Moderator</option>
                </select>
                <TextMuted>Account Type</TextMuted>
              </label>
              <Extension>
                <div className="helper-text">
                  <ul>
                    <li>
                      <strong>Basic: </strong> Submit questions and play
                      published games.
                    </li>
                    <li>
                      <strong>Moderator: </strong> Basic access, plus game
                      creation.
                    </li>
                  </ul>
                </div>
              </Extension>
              <Button primary className="block" type="submit">
                Create Account
              </Button>
              <Extension>
                <Link
                  href={{
                    pathname: 'signin',
                  }}
                >
                  <Button tertiary>Have An Account? Sign In</Button>
                </Link>
              </Extension>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default withUser(Signup);
export { SIGNUP_MUTATION };
