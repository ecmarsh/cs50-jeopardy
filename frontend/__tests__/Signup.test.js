import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import Signup, { SIGNUP_MUTATION } from '../components/Signup';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';
import Router from 'next/router';

function mockInput(wrapper, name, value) {
  wrapper.find(`input[name="${name}"]`).simulate('change', {
    target: { name, value },
  });
}

const me = fakeUser();

const signupMock = {
  request: {
    query: SIGNUP_MUTATION,
    variables: {
      name: me.name,
      email: me.email,
      password: 'fakePassword',
      access: 'moderator',
    },
  },
  result: {
    data: {
      signup: {
        __typename: 'User',
        id: me.id,
        email: me.email,
        name: me.name,
      },
    },
  },
};

const refetchMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...me,
        __typename: 'User',
      },
    },
  },
};

const noUserMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        __typename: 'User',
        id: '',
        email: '',
        name: '',
        permissions: [],
      },
    },
  },
};

describe('<Signup />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      // Pass empty user since it checks for that first
      <MockedProvider mocks={[noUserMock]}>
        <Signup />
      </MockedProvider>
    );
    expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
  });

  it('calls signup mutation and logs user in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={[signupMock, noUserMock, refetchMock]}>
        <Signup />
      </MockedProvider>
    );

    mockInput(wrapper, 'name', me.name);
    mockInput(wrapper, 'email', me.email);
    mockInput(wrapper, 'password', 'fakePassword');
    wrapper.find('select').simulate('change', {
      target: { name: 'access', value: 'moderator' },
    });

    // Mock router
    Router.router = { push: jest.fn() };

    // Submit the form
    wrapper.find('form').simulate('submit');
    // Wait for res
    await wait(50);
    wrapper.update();
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith({ pathname: '/' });

    wrapper.update();
    expect(wrapper.text()).toContain('Looks like you are already a user');

    let apolloClient = wrapper.find('ApolloProvider').props().client;
    const user = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(user.data.me).toMatchObject(me);
  });
});
