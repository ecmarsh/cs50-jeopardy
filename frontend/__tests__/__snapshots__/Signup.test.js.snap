// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> renders and matches snapshot 1`] = `
<form
  className="Form-qsq7s5-0 bLtDzK"
  method="POST"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <h2>
    Create Account
  </h2>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="name"
    >
      <input
        id="name"
        minLength="2"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        required={true}
        type="text"
        value=""
      />
      <Utilities__TextMuted>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Utilities__TextMuted-s11g5i-2",
                "isStatic": false,
                "lastClassName": "bdrBqE",
                "rules": Array [
                  "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                ],
              },
              "displayName": "Utilities__TextMuted",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Utilities__TextMuted-s11g5i-2",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="Utilities__TextMuted-s11g5i-2 bdrBqE"
          >
            Name
          </span>
        </StyledComponent>
      </Utilities__TextMuted>
    </label>
    <label
      htmlFor="email"
    >
      <input
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        type="email"
        validate="true"
        value=""
      />
      <Utilities__TextMuted>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Utilities__TextMuted-s11g5i-2",
                "isStatic": false,
                "lastClassName": "bdrBqE",
                "rules": Array [
                  "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                ],
              },
              "displayName": "Utilities__TextMuted",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Utilities__TextMuted-s11g5i-2",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="Utilities__TextMuted-s11g5i-2 bdrBqE"
          >
            Email
          </span>
        </StyledComponent>
      </Utilities__TextMuted>
    </label>
    <label
      htmlFor="password"
    >
      <input
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        required={true}
        type="password"
        value=""
      />
      <Utilities__TextMuted>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Utilities__TextMuted-s11g5i-2",
                "isStatic": false,
                "lastClassName": "bdrBqE",
                "rules": Array [
                  "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                ],
              },
              "displayName": "Utilities__TextMuted",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Utilities__TextMuted-s11g5i-2",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="Utilities__TextMuted-s11g5i-2 bdrBqE"
          >
            Password
          </span>
        </StyledComponent>
      </Utilities__TextMuted>
    </label>
    <label
      htmlFor="access"
    >
      <select
        defaultValue="basic"
        id="access"
        name="access"
        onChange={[Function]}
        required={true}
        style={
          Object {
            "minHeight": "3.5rem",
            "minWidth": "27rem",
          }
        }
      >
        <option
          value="basic"
        >
          Basic
        </option>
        <option
          value="moderator"
        >
          Moderator
        </option>
      </select>
      <Utilities__TextMuted>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Utilities__TextMuted-s11g5i-2",
                "isStatic": false,
                "lastClassName": "bdrBqE",
                "rules": Array [
                  "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                ],
              },
              "displayName": "Utilities__TextMuted",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Utilities__TextMuted-s11g5i-2",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="Utilities__TextMuted-s11g5i-2 bdrBqE"
          >
            Account Type
          </span>
        </StyledComponent>
      </Utilities__TextMuted>
    </label>
    <Form__Extension>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "Form__Extension-qsq7s5-1",
              "isStatic": false,
              "lastClassName": "enwYgf",
              "rules": Array [
                "display:flex;justify-content:space-between;min-width:22rem;max-width:30rem;margin:0 auto;.helper-text{font-size:1.2rem;text-align:left;transform:translateX(10%);color:",
                [Function],
                ";}",
              ],
            },
            "displayName": "Form__Extension",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "Form__Extension-qsq7s5-1",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="Form__Extension-qsq7s5-1 enwYgf"
        >
          <div
            className="helper-text"
          >
            <ul>
              <li>
                <strong>
                  Basic: 
                </strong>
                 Submit questions and play published games.
              </li>
              <li>
                <strong>
                  Moderator: 
                </strong>
                 Basic access, plus game creation.
              </li>
            </ul>
          </div>
        </div>
      </StyledComponent>
    </Form__Extension>
    <Button
      className="block"
      primary={true}
      type="submit"
    >
      <StyledComponent
        className="block"
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "Button-sc-1f1w3gq-0",
              "isStatic": false,
              "lastClassName": "kpEvwZ",
              "rules": Array [
                "font-family:",
                [Function],
                ";color:white;border-radius:0.5rem;text-transform:uppercase;letter-spacing:1px;transition:.2s ease;&:hover{cursor:pointer;}&:focus{cursor:pointer;outline:",
                [Function],
                " auto 3px;}",
                [Function],
                " ",
                [Function],
                " ",
                [Function],
                " ",
                [Function],
                " ",
                [Function],
                " &.button--sound{font-weight:bolder;text-decoration:line-through;&::after{content:\\"❌\\";padding-left:3px;}",
                [Function],
                "}",
                [Function],
              ],
            },
            "displayName": "Button",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "Button-sc-1f1w3gq-0",
            "target": "button",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        primary={true}
        type="submit"
      >
        <button
          className="block Button-sc-1f1w3gq-0 jSaPjy"
          type="submit"
        >
          Create Account
        </button>
      </StyledComponent>
    </Button>
    <Form__Extension>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "Form__Extension-qsq7s5-1",
              "isStatic": false,
              "lastClassName": "enwYgf",
              "rules": Array [
                "display:flex;justify-content:space-between;min-width:22rem;max-width:30rem;margin:0 auto;.helper-text{font-size:1.2rem;text-align:left;transform:translateX(10%);color:",
                [Function],
                ";}",
              ],
            },
            "displayName": "Form__Extension",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "Form__Extension-qsq7s5-1",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="Form__Extension-qsq7s5-1 enwYgf"
        >
          <Link
            href={
              Object {
                "pathname": "signin",
              }
            }
          >
            <Button
              onClick={[Function]}
              tertiary={true}
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "Button-sc-1f1w3gq-0",
                      "isStatic": false,
                      "lastClassName": "kpEvwZ",
                      "rules": Array [
                        "font-family:",
                        [Function],
                        ";color:white;border-radius:0.5rem;text-transform:uppercase;letter-spacing:1px;transition:.2s ease;&:hover{cursor:pointer;}&:focus{cursor:pointer;outline:",
                        [Function],
                        " auto 3px;}",
                        [Function],
                        " ",
                        [Function],
                        " ",
                        [Function],
                        " ",
                        [Function],
                        " ",
                        [Function],
                        " &.button--sound{font-weight:bolder;text-decoration:line-through;&::after{content:\\"❌\\";padding-left:3px;}",
                        [Function],
                        "}",
                        [Function],
                      ],
                    },
                    "displayName": "Button",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "Button-sc-1f1w3gq-0",
                    "target": "button",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                onClick={[Function]}
                tertiary={true}
              >
                <button
                  className="Button-sc-1f1w3gq-0 kpEvwZ"
                  onClick={[Function]}
                >
                  Have An Account? Sign In
                </button>
              </StyledComponent>
            </Button>
          </Link>
        </div>
      </StyledComponent>
    </Form__Extension>
  </fieldset>
</form>
`;
